libazureinit/src/health.rs

```diff
@@ -0,0 +1,418 @@
+            reason: reason.into(),
+            agent: format!("Azure-Init/{}", env!("CARGO_PKG_VERSION")),
+            documentation_url: "https://aka.ms/linuxprovisioningerror",
+            vm_id: vm_id.into(),
+            timestamp: Utc::now(),
+            pps_type: Some("None".to_string()),
+            extra: HashMap::new(),
+        }
+    }
+
+    fn with_extra(mut self, key: &str, value: &str) -> Self {
+        self.extra.insert(key.to_string(), value.to_string());
+        self
+    }
+}
+
+/// Report that provisioning is ready.
+pub async fn report_ready(config: &Config) -> Result<(), Error> {
+    tracing::info!("Reporting provisioning complete");
+    _report(ProvisioningState::Ready, None, None, &config.wireserver).await
+}
+
+/// Report that provisioning has failed.
+pub async fn report_failure(
+    message: &str,
+    config: &Config,
+    vm_id: &str,
+) -> Result<(), Error> {
+    let error = ReportableError::new(message, vm_id.to_string())
+        .with_extra("component", "provisioning");
+
+    tracing::info!(
+        reason = %error.reason,
+        vm_id = %error.vm_id,
+        "Reporting provisioning failure"
+    );
+
+    let body = serde_json::to_string(&error)?;
+
+    _report(
+        ProvisioningState::NotReady,
+        Some(ProvisioningSubStatus::ProvisioningFailed),
+        Some(&body),
+        &config.wireserver,
+    )
+    .await
+}
+
+/// Report that provisioning is in progress.
+pub async fn report_in_progress(
+    message: &str,
+    config: &Config,
+) -> Result<(), Error> {
+    _report(
+        ProvisioningState::InProgress,
+        Some(ProvisioningSubStatus::ProvisioningInProgress),
+        Some(message),
+        &config.wireserver,
+    )
+    .await
+}
+
+/// Internal helper: all of the header setup, JSON‐body construction, retry loop, etc.
+#[instrument(err, skip_all)]
+async fn _report(
+    state: ProvisioningState,
+    substatus: Option<ProvisioningSubStatus>,
+    description: Option<&str>,
+    cfg: &crate::config::Wireserver,
+) -> Result<(), Error> {
+    let body = if let Some(sub) = substatus {
+        json!({
+            "state": state.to_string(),
+            "details": {
+                "subStatus": sub.to_string(),
+                "description": description
+            }
+        })
+        .to_string()
+    } else {
+        json!({ "state": state.to_string() }).to_string()
+    };
+
+    tracing::debug!(body=%body, "Built provisioning-health JSON");
+
+    let version = env!("CARGO_PKG_VERSION");
+    let mut headers = HeaderMap::new();
+    headers.insert(
+        USER_AGENT,
+        HeaderValue::from_str(&format!("azure-init v{}", version)).unwrap(),
+    );
+    headers.insert(
+        "x-ms-guest-agent-name",
+        HeaderValue::from_str(&format!("azure-init/{}", version)).unwrap(),
+    );
+    headers
+        .insert("content-type", HeaderValue::from_static("application/json"));
+
+    tracing::info!(?headers, "Prepared HTTP headers");
```

Comment:
IMO info is a little high for this. I'd recommend debug or even trace level since the header content won't be particularly interesting in most cases.